src/App.tsx

```diff
@@ -1,223 +1,236 @@
+        });
+    }
+  }, []);
+
+  const uncompletedTodosLength = useMemo(() => (
+    todos.filter(todo => !todo.completed).length
+  ), [todos]);
+
+  const completedTodosLength = useMemo(() => (
+    todos.filter(todo => todo.completed).length
+  ), [todos]);
+
+  const visibleTodos = useMemo(() => {
+    return filterTodos(todos, complitedFilter);
+  }, [complitedFilter, todos]);
+
+  const isAllTodosCompleted = useMemo(() => (
+    todos.every(todo => todo.completed)
+  ), [todos]);
+
+  const deleteTodo = useCallback(async (todoId: number) => {
+    try {
+      setUpdatingTodos(prev => [...prev, todoId]);
+      const deleteResponse = await deleteTodoById(todoId);
+
+      setTodos(prevTodos => (
+        prevTodos.filter(todo => todo.id !== todoId)
+      ));
+
+      return deleteResponse;
+    } catch (deleteError) {
+      return setErrorMessage('Unable to delete a todo');
+    } finally {
+      setUpdatingTodos([]);
+    }
   }, []);
 
+  const clearCompletedTodos = useCallback(() => {
+    todos.forEach(todo => {
+      if (todo.completed) {
+        deleteTodo(todo.id);
+      }
+    });
+  }, [todos]);
+
+  const addTodo = useCallback(
+    async (title: string) => {
+      try {
+        if (!title.trim()) {
+          setErrorMessage('Title can\'t be empty');
+
+          return;
+        }
+
+        setIsNewTodoLoading(true);
+
+        if (user) {
+          setTempTodo({
+            id: 0,
+            userId: user.id,
+            title,
+            completed: false,
+          });
+
+          const newTodo = await createTodo({
+            userId: user.id,
+            title,
+            completed: false,
+          });
+
+          setTodos(prevTodos => [...prevTodos, newTodo]);
+          setTempTodo(null);
+        }
+      } catch (addTodoError) {
+        setErrorMessage('Unable to add a todo');
+      } finally {
+        setIsNewTodoLoading(false);
+      }
+    }, [todos, user],
+  );
+
+  const toggleTodoStatus = useCallback(async (
+    todoId: number,
+    updatedTodoStatus: boolean,
+  ) => {
+    try {
+      setUpdatingTodos(prev => [...prev, todoId]);
+      const updateResponse = await updateTodoById(
+        { completed: updatedTodoStatus }, todoId,
+      );
+
+      setTodos(prevTodos => (
+        prevTodos.map(todo => {
+          return todo.id === todoId
+            ? updateResponse
+            : todo;
+        })
+      ));
+
+      return updateResponse;
+    } catch (updateError) {
+      return setErrorMessage('Unable to update a todo');
+    } finally {
+      setUpdatingTodos([]);
+    }
+  }, [todos]);
+
+  const toggleAllTodos = useCallback(() => {
+    const todosChangingStatus = todos.filter(todo => (
+      todo.completed === isAllTodosCompleted
+    ));
+
+    todosChangingStatus.forEach(todo => {
+      toggleTodoStatus(todo.id, !isAllTodosCompleted);
+    });
+  }, [todos, isAllTodosCompleted]);
+
+  const updateTodoTitle = useCallback(async (
```

Comment:
You could write one `update` function and use it inside `updateTodoTitle` and `toggleTodoStatus`

Reply:
done

Reply:
done